### PR TITLE
infra: better type Map

### PR DESCRIPTION
**Pull request by stephanwlee**
When you create a Map without values, TypeScript has to infer the type
and, sometimes, it is imperfect and types cannot unsafe. This change
forces valueless Map constructor to specify the types so we rely less on
the inference.

To emulate the internal enforcement, we are making a necessary changes
to `tf_ts_library` but do note that it is imperfect and sometimes misses the
cases. Also do note that it somehow leaks its way into targets with
`strict_check = False` which is unintended but bearable.

Do note that this requires sync changes.
Test CL with necessary change patched: cl/350867815

### CODE REVIEW

File: tensorboard/defs/strict_type_check.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an 'AS IS' BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Type override of lib.es2015.collection.d.ts.
+// Make sure value is not inferred and take anything.
```
Comment (psybuzz) [inline tensorboard/defs/strict_type_check.d.ts:17]: nit: 'Make sure key/value types are not inferred as `any`.'?

**Review by psybuzz — COMMENTED** [review]
Nice!  Internally, is the introduction of this new rule something that we are driving ourselves?

**Comment by stephanwlee** [comment]
No, we are not driving it. For context, PTAL cl/350606198.

**Review by psybuzz — APPROVED** [review]
Thanks for the internal CL context, this PR lgtm

**Pull request merged**